src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/TypeMapLazyDictionary.cs

```diff
@@ -0,0 +1,310 @@
+            delegate* unmanaged<CallbackContext*, ProcessAttributesCallbackArg*, void> newProxyTypeEntry,
+            CallbackContext* context);
+
+        private static TypeName CreateNewTypeName(string typeName, string assemblyName)
+        {
+            AssemblyNameInfo asmNameInfo = AssemblyNameInfo.Parse(assemblyName);
+            TypeName parsedType = new TypeName(typeName, asmNameInfo);
+            Debug.Assert(parsedType.AssemblyName != null);
+            return parsedType;
+        }
+
+        [UnmanagedCallersOnly]
+        private static unsafe void NewExternalTypeEntry(CallbackContext* context, ProcessAttributesCallbackArg* arg)
+        {
+            Debug.Assert(context != null);
+            Debug.Assert(arg != null);
+            Debug.Assert(arg->Utf8String1 != null);
+            Debug.Assert(arg->Utf8String2 != null);
+
+            string externalTypeName = new((sbyte*)arg->Utf8String1, 0, arg->StringLen1, Encoding.UTF8);
+            string targetType = new((sbyte*)arg->Utf8String2, 0, arg->StringLen2, Encoding.UTF8);
+
+            TypeName parsedTarget = TypeNameParser.Parse(targetType, throwOnError: true)!;
+            if (parsedTarget.AssemblyName is null)
+            {
+                // The assembly name is not included in the type name, so use the fallback assembly name.
+                Debug.Assert(arg->Utf8String3 != null);
+                string fallbackAssemblyName = new((sbyte*)arg->Utf8String3, 0, arg->StringLen3, Encoding.UTF8);
+                parsedTarget = CreateNewTypeName(targetType, fallbackAssemblyName);
+            }
+            context->ExternalTypeMap.Add(externalTypeName, parsedTarget);
+        }
+
+        [UnmanagedCallersOnly]
+        private static unsafe void NewProxyTypeEntry(CallbackContext* context, ProcessAttributesCallbackArg* arg)
+        {
+            Debug.Assert(context != null);
+            Debug.Assert(arg != null);
+            Debug.Assert(arg->Utf8String1 != null);
+            Debug.Assert(arg->Utf8String2 != null);
+
+            string sourceType = new((sbyte*)arg->Utf8String1, 0, arg->StringLen1, Encoding.UTF8);
+            string proxyType = new((sbyte*)arg->Utf8String2, 0, arg->StringLen2, Encoding.UTF8);
+
+            TypeName parsedSource = TypeNameParser.Parse(sourceType, throwOnError: true)!;
+            TypeName parsedProxy = TypeNameParser.Parse(proxyType, throwOnError: true)!;
+            if (parsedSource.AssemblyName is null || parsedProxy.AssemblyName is null)
+            {
+                // The assembly name is not included in the type name, so use the fallback assembly name.
+                Debug.Assert(arg->Utf8String3 != null);
+                string fallbackAssemblyName = new((sbyte*)arg->Utf8String3, 0, arg->StringLen3, Encoding.UTF8);
+                if (parsedSource.AssemblyName is null)
+                {
+                    parsedSource = CreateNewTypeName(sourceType, fallbackAssemblyName);
```

Comment:
There is `WithAssemblyName` API that you can use to "add" assembly name to a type name. This API is under `SYSTEM_REFLECTION_METADATA` ifdef, but you should be avoid to relax that or create a similar API for . It will avoid reparsing of the whole string just to attach an assembly name.

Reply:
Done.